flang/lib/Optimizer/HLFIR/Transforms/SimplifyHLFIRIntrinsics.cpp

```diff
@@ -375,6 +478,24 @@ class SimplifyHLFIRIntrinsics
       }
       return true;
     });
+    target.addDynamicallyLegalOp<hlfir::CShiftOp>([](hlfir::CShiftOp cshift) {
```

Comment:
I think it would be easier to turn this pass into a greedy conversion, and check all the conditions in the patterns themselves. If there are no objections, I will clean it up in a separate NFC patch.